src/library/Box/Tools.php

```diff
@@ -562,4 +562,25 @@ public function validateAndSanitizeEmail($email, $throw = true){
 
         return $email;
     }
+    /**
+     *  Remove empty values in a array. Issue is caused by the fact forms need to contain at least 1 value to be submitted...
+     *  @param array
+     *  return null or array;
+    */
+    public function removeEmptyValues($tobecleaned){
```

Comment:
Realized this AM, you are re-inventing a function that already exists. Simply use `array_filter`.
Example: `$upgrades = array_filter($upgrades);` will remove all empty entries of the array.

From the PHP docs, the 2nd param is the callback function to use and not supplying it then has exactly the behavior we want: 
> If no callback is supplied, all empty entries of array will be removed. See [empty()](https://www.php.net/manual/en/function.empty.php) for how PHP defines empty in this case.

https://www.php.net/manual/en/function.array-filter.php

Reply:
array_values(array_filter($array)))); is the trick

Reply:
You turned it into json, and it outputted json. 

But if you look, it stripped out the unwanted array values as it should have.

I fail to see what the issue is besides you expecting it to return an array when you are using a function to turn it into a json string

Reply:
Using <?php array_filter($data['upgrades']);?> and json_encode the result causes Fossbilling to crash. 

> An exception has been thrown during the rendering of a template ("SQLSTATE[HY093]: Invalid parameter number: parameter was not defined").

It doesn't allow the array keys to be provided when you encode with json_encode();

Reply:
The post data is JSON, you need to use json_decode to convert it FROM JSON and into an array.
I do not understand what the confusion is

Reply:
The code has been updated and should work...

Reply:
After a further look into the code, I see the api data is converted to an array before it's sent to this function, I had that mixed up